src/cable/server.cr

```diff
@@ -148,17 +129,23 @@ module Cable
       end
     end
 
-    private def subscribe
+    private def subscribe_to_internal
       spawn(name: "Cable::Server - subscribe") do
-        redis_subscribe.subscribe("_internal") do |on|
-          on.message do |channel, message|
-            if channel == "_internal" && message == "debug"
-              debug
-            else
-              fiber_channel.send({channel, message})
-              Cable::Logger.debug { "Cable::Server#subscribe channel:#{channel} message:#{message}" }
+        begin
+          redis_subscribe.subscribe("_internal") do |subscription|
+            subscription.on_message do |channel, message|
+              if channel == "_internal" && message == "ping"
+                Cable::Logger.debug { "Cable::Server#subscribe channel:#{channel} message:PONG" }
+              elsif channel == "_internal" && message == "debug"
+                debug
+              else
+                fiber_channel.send({channel, message})
+                Cable::Logger.debug { "Cable::Server#subscribe channel:#{channel} message:#{message}" }
+              end
             end
           end
+        rescue e : IO::Error
+          # why is redis_subscribe.@socket.closed? here??
         end
```

Comment:
Between each spec we call `Cable.restart`. After the first spec runs, this block here ends up throwing an `IO::Error`. It seems when we try to run `redis_subscribe.subscribe("_internal")`, the `redis_subscribe.@socket` is already `closed?`. It should be a new instance with a new socket, but I'm not sure why it's not... For now, we can ignore it

Reply:
As mentioned in https://github.com/jgaskins/redis/issues/12, I played around with this trying to understand what's going on. I added a bunch of `puts` statements all over the place to see the order of operations.

<details><summary>Changes made to code in this PR and monkeypatches to Redis</summary>

```crystal
# module Cable
  def self.restart
    if current_server = @@server
      puts "Shutting down old server..."
      current_server.shutdown
    end
    puts "Spinning up new server..."
    @@server = Server.new
  end

# class Cable::Server
    # Use the pooled connections
    getter redis_publish = begin
      puts "Connecting publisher"
      Redis::Client.new(URI.parse(Cable.settings.url))
    end
    # Use a single connection
    getter redis_subscribe = begin
      puts "Connecting subscriber"
      Redis::Connection.new(URI.parse(Cable.settings.url))
    end

# in Redis::Connection#initialize
      pp connecting: @uri.to_s
      @writer = Writer.new(IO::MultiWriter.new(STDERR, socket))
```

</details>

There may also be a `Fiber.yield` in there to try to give the spawned fibers a chance to run:

---

Here's the output for an `IO::Error` chosen arbitrarily:

```text
Shutting down old server...
*1
$11
unsubscribe
Spinning up new server...
Connecting publisher
{connecting: "redis://localhost:6379"}
*Unhandled exception in spawn(name: Cable::Server - subscribe): Closed stream (IO::Error)
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io.cr:118:5 in 'check_open'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io/buffered.cr:129:5 in 'write'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io/multi_writer.cr:37:30 in 'write'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io.cr:877:5 in 'write_byte'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/char.cr:897:9 in 'to_s'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/io.cr:174:5 in '<<'
  from lib/redis/src/writer.cr:14:7 in 'encode'
  from lib/redis/src/connection.cr:136:7 in 'subscribe'
  from lib/redis/src/connection.cr:334:9 in 'call'
  from lib/redis/src/connection.cr:119:5 in '->'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/fiber.cr:146:11 in 'run'
  from /opt/homebrew/Cellar/crystal/1.5.0/share/crystal/src/fiber.cr:98:34 in '->'
Connecting subscriber
{connecting: "redis://localhost:6379"}
Initializing server...
*2
$9
subscribe
$9
_internal
```

It looks like things are running out of order? That doesn't make any sense, though.

Reply:
Oh... Is it trying to subscribe to the _internal before the subscribe has even been connected? :thinking:

Reply:
That's what I thought at first, and I was like "Crystal y u like dis?" and then [this realization hit me](https://github.com/jgaskins/redis/issues/12#issuecomment-1186628415)